File: python/15-3Sum.py
# region Sliding Window
# Space O(n)
# Time O(n)
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        retVal = []

        for i in range(len(nums) - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            left, right = i + 1, len(nums) - 1

            while left < right:
                total = nums[i] + nums[left] + nums[right]

                if total == 0:
                    retVal.append([nums[i], nums[left], nums[right]])
                    left += 1
                    right -= 1
                    while nums[left] == nums[left - 1] and left < right:
                        left += 1
                    while nums[right] == nums[right + 1] and left < right:
                        right -= 1
                elif total < 0:
                    left += 1
                else:
                    right -= 1
        return retVal
# endregion

File: c#/15-3Sum.cs
public class Solution
{
    public List<List<int>> ThreeSum(int[] nums)
    {
        var res = new List<List<int>>();
        Array.Sort(nums);

        for (int i = 0; i < nums.Length; i++)
        {
            if (nums[i] > 0)
                break;
            if (i > 0 && nums[i] == nums[i - 1])
                continue;

            int l = i + 1,
                r = nums.Length - 1;
            while (l < r)
            {
                int curSum = nums[i] + nums[l] + nums[r];

                if (curSum > 0)
                {
                    r--;
                }
                else if (curSum < 0)
                {
                    l++;
                }
                else
                {
                    res.Add(new List<int> { nums[i], nums[l], nums[r] });
                    l++;
                    r--;
                    while (l < r && nums[l] == nums[l - 1])
                    {
                        l++;
                    }
                }
            }
        }
        return res;
    }
}
